File: ThisIsCodingTest/이것이코딩테스트다_12장_문자열압축.js
const addWord = (word, cnt) => {
  return cnt === 1 ? word : cnt.toString() + word;
}

const getIdx = (i, length, s) => {
  let idx = i + length;

  if (idx > s.length) {
    idx = s.length;
  }

  return idx;
}

const solution = (s) => {
  if (s.length === 1) {
    return 1;
  }

  let answer = Infinity;
  let length = Math.floor(s.length / 2);;

  while (length > 0) {
    let word = s.slice(0, length);
    let words = '';
    let cnt = 1;

    for (let i = length; i < s.length; i += length) {
      let idx = getIdx(i, length, s);

      const targetWord = s.slice(i, idx);

      if (word === targetWord) {
        cnt += 1;
        continue;
      }

      words += addWord(word, cnt);
      word = targetWord;
      cnt = 1;
    }

    words += addWord(word, cnt);;
    answer = Math.min(answer, words.length);
    length -= 1;
  }

  return answer;
}

console.log(solution('a'));
console.log(solution('aabbaccc'));
console.log(solution('ababcdcdababcdcd'));
console.log(solution('abcabcdede'));
console.log(solution('abcabcabcabcdededededede'));
console.log(solution('xababcdcdababcdcd'));
// index 11 ~ 15
// 11 + 6 => 17

// 짝수
// aabb accc
// 8 -> 4

// 홀수
// aa bba
// 5 -> 2

File: this-is-coding-test/이것이코딩테스트다_12장_문자열압축.py
def solution(s):
    arr_s = list(s)
    # 글자수 보다 커질 수 없음
    min_length = len(s)
    start = 1

    while start <= len(s) // 2:
        temp_str = ''
        temp_length = 1
        target = arr_s[:start]
        for i in range(start, len(s), start):
            if target == arr_s[i: i + start]:
                temp_length += 1
            else:
                if temp_length > 1:
                    temp_str += str(temp_length) + ''.join(target)
                else:
                    temp_str += ''.join(target)
                temp_length = 1
                target = arr_s[i: i + start]

        # for 문 끝나고 마지막에 남은 문자열 처리
        if temp_length > 1:
            temp_str += str(temp_length) + ''.join(target)
        else:
            temp_str += ''.join(target)

        min_length = min(min_length, len(temp_str))
        start += 1

    return min_length


print(solution("aabbaccc"))
print(solution("ababcdcdababcdcd"))
print(solution("abcabcdede"))
print(solution("abcabcabcabcdededededede"))
print(solution("xababcdcdababcdcd"))

print(solution("a"))

# 문제를 이해하는데 오래걸렸다
# 이 문제는 무조건 특정 갯수대로 잘라나가야 한다
